protocol/eth-shared-lockbox.md

```diff
@@ -0,0 +1,164 @@
+    }
+    // Update the dependency set
+    dependencySet.push(chainId);
+    isRegistered[chainId] = true;
+
+    emit ChainAdded(chainId);
+}
+    
+```
+
+For `removeChain`, the process would follows the same logic.
+
+Note that, under the specified flow, the dependency set consistently maintains the form of a [complete graph](https://en.wikipedia.org/wiki/Complete_graph) at all times.
+
+### `addDependencies` and `removeDependencies` in `SystemConfigInterop` and `L1BlockInterop`
+
+The purpose of these functions is to extend the current `SystemConfigInterop` implementation by allowing multiple `chainId` values to be added in a single call by passing them as an array. There are a few ways to pass these values through `setConfig`: one is making a single call and triggering the `TransactionDeposited` event, or by looping through the current `addDependency`/`removeDependency` functions.
+
+With this optimization, the `DependencySetManager` can simplify the number of calls, as follows:
+
+```solidity
+function addChain(uint256 chainId) external onlyOwner {
+  // ...
+	if (dependencySet.length > 0) {
+				// If we can input an array chainId values thorugh addDependencies function
+				ISystemConfigInterop(newChainSystemConfig).addDependencies(dependencySet);
+	}		
+	// ...
+```
+
+### `SharedLockbox` implementation
+
+A minimal set of functions should include:
+
+- `lockETH`: Accepts ETH from the `depositTransaction` originating from a valid `OptimismPortal2`.
+- `unlockETH`: Releases ETH from the `finalizeWithdrawalTransaction` originating from a valid `OptimismPortal2`.
+
+Access control for `lockETH` and `unlockETH` is validated against the list reated by `registerChain`.
+
+### `OptimismPortal2` upgrade process
+
+A one-time L1 liquidity migration is required for each approved chain. By using an intermediate contract during the upgrade, all ETH held by the `OptimismPortal` can be transferred to the `SharedLockbox` within the `initialize` function. After this migration, the `OptimismPortal` is upgraded to the new version with the desired functionality.
+
+Importantly, the entire upgrade process—including migrating the ETH to the `SharedLockbox` and updating the `OptimismPortal` to the latest version—can be executed in a single transaction. This approach ensures a migration without the necessity of maintaining a persistent migration function in the final contract.
+
+Only chains registered in the `DependencySetManager` should be approved to join the `SharedLockbox`. Once approved, the `SharedLockbox` address is added in the `constructor` and `initialize` functions for each `OptimismPortal2`.
+
+## Impact
+
+The following components require an audit of the new and modified contracts: `OptimismPortal2`, `SharedLockbox`, `SystemConfigInterop`, `L1BlockInterop`, and `DependencySetManager`. This also includes the scripts necessary to perform the migration.
+
+# Alternatives Considered
```

Comment:
Related to some of my questions about (why do we need dependency set info in the system config and the DependencySetManager, and do we really need the DependencySetManager), I'd like to see more alternatives and justification for this specific lockbox design, compared to other lockbox designs. 

An example: Have we considered deprecating the portal, and having the lockbox emit deposit events with a chain ID? The obvious downside is now this requires changes to the derivation path, but it does mean we can get rid of an entire contract which is nice

Reply:
Deprecating the portal in favor of developing a new SharedPortal sounds like a good long-term goal, but it would be far beyond the scope of the ETH Lockbox feature. Creating this new SharedPortal looks like a massive and highly sensitive project that would need debate and coordination across multiple teams to bring it to life. We think this is a valid but complex question that could significantly alter the direction of this task. We're interested to hear what others think about it.

Reply:
I would prefer to not add more derivation changes. We have considered a "beacon" style deposit contract, this could be on the roadmap in the future but this feature doesn't make or break the design. I am in favor of the minimal diff that maintains the portal contracts and simply holds its ether in a shared lockbox

Reply:
We do need a dependency set manager that orchestrates the dependency set of the entire cluster, unless we do have a shared system config